src/lib/snarky_blake2/snarky_blake2.ml

```diff
@@ -0,0 +1,268 @@
+open Core_kernel
+
+module Make (Impl : Snarky.Snark_intf.S) = struct
+  open Impl
+  open Let_syntax
+  module UInt32 = Uint32.Make (Impl)
+
+  let r1 = 16
+
+  let r2 = 12
+
+  let r3 = 8
+
+  let r4 = 7
+
+  let ( := ) (v, i) t =
+    let%map x = t in
+    v.(i) <- x
+
+  let mixing_g v a b c d x y =
+    let ( := ) i t = (v, i) := t in
+    let open UInt32 in
+    let xorrot t1 t2 k = xor t1 t2 >>| Fn.flip UInt32.rotr k in
+    let%bind () = a := sum [v.(a); v.(b); x] in
+    let%bind () = d := xorrot v.(d) v.(a) r1 in
+    let%bind () = c := sum [v.(c); v.(d)] in
+    let%bind () = b := xorrot v.(b) v.(c) r2 in
+    let%bind () = a := sum [v.(a); v.(b); y] in
+    let%bind () = d := xorrot v.(d) v.(a) r3 in
+    let%bind () = c := sum [v.(c); v.(d)] in
+    let%bind () = b := xorrot v.(b) v.(c) r4 in
+    return ()
+
+  let iv =
+    Array.map
+      ~f:(Fn.compose UInt32.constant UInt32.Unchecked.of_int)
+      [| 0x6A09E667
+       ; 0xBB67AE85
+       ; 0x3C6EF372
+       ; 0xA54FF53A
+       ; 0x510E527F
+       ; 0x9B05688C
+       ; 0x1F83D9AB
+       ; 0x5BE0CD19 |]
+
+  let splitu64 u =
+    let open Unsigned.UInt64 in
+    let open Infix in
+    let uint32 = Fn.compose UInt32.Unchecked.of_int to_int in
+    let low = uint32 (u land of_int 0xffffffff) in
+    let high = uint32 (u lsr 32) in
+    (low, high)
+
+  let for_ n ~f =
+    let rec go i =
+      if i = n then return ()
+      else
+        let%bind () = f i in
+        go (i + 1)
+    in
+    go 0
+
+  let sigma =
+    [| [|0; 1; 2; 3; 4; 5; 6; 7; 8; 9; 10; 11; 12; 13; 14; 15|]
+     ; [|14; 10; 4; 8; 9; 15; 13; 6; 1; 12; 0; 2; 11; 7; 5; 3|]
+     ; [|11; 8; 12; 0; 5; 2; 15; 13; 10; 14; 3; 6; 7; 1; 9; 4|]
+     ; [|7; 9; 3; 1; 13; 12; 11; 14; 2; 6; 5; 10; 4; 0; 15; 8|]
+     ; [|9; 0; 5; 7; 2; 4; 10; 15; 14; 1; 11; 12; 6; 8; 3; 13|]
+     ; [|2; 12; 6; 10; 0; 11; 8; 3; 4; 13; 7; 5; 15; 14; 1; 9|]
+     ; [|12; 5; 1; 15; 14; 13; 4; 10; 0; 7; 6; 3; 9; 2; 8; 11|]
+     ; [|13; 11; 7; 14; 12; 1; 3; 9; 5; 0; 15; 4; 8; 6; 2; 10|]
+     ; [|6; 15; 14; 9; 11; 3; 0; 8; 12; 2; 13; 7; 1; 4; 10; 5|]
+     ; [|10; 2; 8; 4; 7; 6; 1; 5; 15; 11; 9; 14; 3; 12; 13; 0|] |]
+
+  let compression h (m : UInt32.t array) t f =
+    assert (Array.length h = 8) ;
+    assert (Array.length m = 16) ;
+    let v = Array.append h iv in
+    let open UInt32 in
+    let tlo, thi = splitu64 t in
+    let%bind () = (v, 12) := xor v.(12) (constant tlo) in
+    let%bind () = (v, 13) := xor v.(13) (constant thi) in
+    let%bind () =
+      if f then (v, 14) := xor v.(14) (constant UInt32.Unchecked.max_int)
```

Comment:
I believe `f` only comes into play while processing the last block of input. `digestif` handles this in the `unsafe_get` function, where it calls `set_lastblock` which sets `f.(0)` to the max int32 (the call to `set_lastnode` which sets `f.(1)` to max int32 as well doesn't happen since `ctx.last_node` is always zero).